### DISCUSSION TITLE
How to handle rpc notifications in Lua

### DISCUSSION THREAD

**Discussion by sakhnik** [neovim/neovim#24291] [Q&A]
Hello, I'm attempting to handle "redraw" notifications after "nvim_ui_attach" in an RPC channel, but can't figure out how to do that.
```Lua
local ch = vim.fn.sockconnect('tcp', 'localhost:44444', {rpc = true})
vim.fn.rpcrequest(ch, "nvim_ui_attach", 80, 25, {ext_linegrid = true})
-- How to setup a callback for "redraw"???
```
I can see with Wireshark that the "redraw" notifications are received by the socket, but they seem to be discarded programmatically. Is there an API call to receive them?

**Comment by justinmk** [comment]
Are you trying to listen to UI events in the same process that produces them? `vim.ui_attach()` may be useful.

**Reply by sakhnik** [reply]
It'd be super useful to be able to get the events in a callback. But it doesn't seem to allow listening to the grid events, just widgets. I'd like to capture the whole UI like it's done in the tests.

**Reply by justinmk** [reply]
But to be clear, this is same-process, right? Surely the approach used in the tests works.

**Reply by sakhnik** [reply]
I'm going to run [nvim-gdb](https://github.com/sakhnik/nvim-gdb/tree/lua) tests with busted in the Neovim process running the plugin. There's an auxiliary script to launch the tests, it's Lua run by a different Neovim process. So I'd be glad to take neovim screenshots either way: from within or outside, just like I'm currently do with a Python client [spy_ui.py](https://github.com/sakhnik/nvim-gdb/blob/lua/test/spy_ui.py).

**Reply by justinmk** [reply]
There is an undocumented `nvim__screenshot` function which may be useful.

**Reply by sakhnik** [reply]
Thanks, this might be right what is required for me!
